src/pyhf/infer/intervals.py

```diff
@@ -2,13 +2,96 @@
 from . import hypotest
 from .. import get_backend
 import numpy as np
+from scipy.optimize import toms748 as _toms748
 
 
 def _interp(x, xp, fp):
     tb, _ = get_backend()
     return tb.astensor(np.interp(x, xp.tolist(), fp.tolist()))
 
 
+def upperlimit_auto(data, model, low, high, level=0.05, atol=2e-12, rtol=1e-15):
```

Comment:
I think it would be great if we can integrate this into `pyhf.infer.upperlimit` via e.g. a `method` arg

`pyhf.infer.upperlimit( ... method = 'toms')` and 
`pyhf.infer.upperlimit( ... method = 'gridscan')`

Reply:
I think `toms748` would work better as a name for the method, since that is easily searchable and will point to the algorithm without needing to read through docstrings (for people who prefer google over that).